R/SEmIRD.R

```diff
@@ -0,0 +1,279 @@
+#' @slot initial_conditions list of values for initial conditions (double).
+#' @slot transmission_parameters list of values for transmission parameters
+#'       (double).
+#'
+#' @import deSolve
+#' @import glue
+#' @import reshape2
+#'
+#' @export SEmIRD
+SEmIRD <- setClass("SEmIRD",
+                  # slots
+                  slots = c(
+                    output_names = "list",
+                    initial_condition_names = "list",
+                    transmission_parameter_names = "list",
+                    initial_conditions = "list",
+                    transmission_parameters = "list"
+                  ),
+                  # prototypes for the slots, automatically set parameter names and
+                  # its data type
+                  prototype = list(
+                    output_names = list("S", "E1", "E2", "E3", "E4", "E5", "I", "R", "D", "Incidence", "Deaths"),
+                    initial_condition_names = list("S0", "E10", "E20", "E30", "E40", "E50", "I0", "R0"),
+                    transmission_parameter_names = list("beta", "kappa", "gamma", "mu"),
+                    initial_conditions = vector(mode = "list", length = 6),
+                    transmission_parameters = vector(mode = "list", length = 4)
+                  )
+)
+
+
+#' @describeIn SEmIRD Retrieves initial conditions of SEmIRD model.
+#'
+#' @param object An object of the class SEmIRD.
+#' 
+#' @export
+setMethod("initial_conditions", "SEmIRD",
+          function(object) object@initial_conditions)
+
+
+#' @describeIn SEmIRD Retrieves transmission parameters of SEmIRD model.
+#'
+#' @param object An object of the class SEmIRD.
+#' 
+#' @export
+setMethod("transmission_parameters", "SEmIRD",
+          function(object) object@transmission_parameters)
+
+# SEmIRD class specific functions
+
+#' @describeIn SEmIRD Setter method for initial conditions (S0, E0, I0 and R0)
+#' of the SEmIRD model.
+#'
+#' All initial conditions must sum up to 1.
+#' If the initial conditions provided to do not sum to 1, an error is thrown.
+#'
+#' @param object an object of the class SEmIRD
+#' @param value (list) list of initial conditions S0, E10, E20, E30, I0, R0.
+#'
+#' @return object of class SEmIRD with initial conditions assigned.
+#' 
+#' @export
+setMethod(
+  "initial_conditions<-", "SEmIRD",
+  function(object, value) {
+    
+    if (mean(names(value) %in% object@initial_condition_names) != 1)
+      stop(paste0("Initial conditions must contain: ",
+                  object@initial_condition_names))
+    init_cond <- value
+    
+    # check inits are numeric
+    for (p in list("S0", "E10", "E20", "E30", "E40", "E50", "I0", "R0")) {
+      if (!is.numeric(init_cond[[p]])) {
+        stop(glue("{p} format must be numeric"))
+      }
+    }
+    
+    # check that the initial conditions are properly normalized
+    if (init_cond$S0 + init_cond$E10 + init_cond$E20 + init_cond$E30 + init_cond$I0 + init_cond$R0 != 1) {
```

Comment:
slightly concerned about floating point != 1: should this be `abs(... - 1) < tol` or similar?

Reply:
Good point. It is a check across all models, and seems we should modify them